crates/biome_configuration/src/editorconfig.rs

```diff
@@ -22,40 +21,194 @@ use crate::{
 };
 
 pub fn parse_str(s: &str) -> Result<EditorConfig, EditorConfigDiagnostic> {
-    // TODO: use serde_path_to_error to emit better parse diagnostics
-    serde_ini::from_str(s).map_err(|err| {
+    EditorConfig::from_str(s).map_err(|err| {
         EditorConfigDiagnostic::ParseFailed(ParseFailedDiagnostic {
-            source: Some(Error::from(IniError::from(err))),
+            source: Some(Error::from(err)),
         })
     })
 }
 
+#[derive(Debug)]
+pub enum EditorConfigErrorKind {
+    StaticStr(&'static str),
+    ParseBoolError(std::str::ParseBoolError),
+    ParseFormatNumberError(ParseFormatNumberError),
+    // A section must end with `]`
+    MissingSectionEnd,
+}
+impl std::fmt::Display for EditorConfigErrorKind {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::StaticStr(s) => f.write_str(s),
+            Self::ParseBoolError(error) => {
+                write!(f, "{}", error)
+            }
+            Self::ParseFormatNumberError(error) => {
+                write!(f, "{}", error)
+            }
+            Self::MissingSectionEnd => f.write_str("a section must be closed with `]`"),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct EditorConfigError {
+    line_number: u32,
+    kind: EditorConfigErrorKind,
+}
+impl biome_diagnostics::Diagnostic for EditorConfigError {
```

Comment:
Now that we own the parsing phase, I wonder if we could compute a `TextRange`, so we can implement the `location()` method. Diagnostics will be better. 

We can optimise it later